File: dotnet/xmscommon/framing/AMQDataBlockEncoder.cs
using System.Collections;
using jpmorgan.mina.common;
using jpmorgan.mina.common.support;
using jpmorgan.mina.filter.codec;
using jpmorgan.mina.filter.codec.demux;
using log4net;

namespace OpenAMQ.Framing
{
    public class AMQDataBlockEncoder : IMessageEncoder
    {
        private static ILog _logger = LogManager.GetLogger(typeof(AMQDataBlockEncoder));

        private Hashtable _messageTypes;
        
        public AMQDataBlockEncoder()
        {
            _messageTypes = new Hashtable();
            _messageTypes[typeof (IEncodableAMQDataBlock)] = 1;
        }


        public Hashtable MessageTypes
        {
            get
            {
                return _messageTypes;
            }
        }

        public void Encode(ISession session, object message, IProtocolEncoderOutput output)
        {
            IDataBlock frame = (IDataBlock) message;
            int frameSize = (int)frame.Size; // TODO: sort out signed/unsigned
            ByteBuffer buffer = ByteBuffer.Allocate(frameSize);
            frame.WritePayload(buffer);
            
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("Encoded frame byte-buffer is '" + ByteBufferHexDumper.GetHexDump(buffer) + "'");
            }
            buffer.Flip();
            output.Write(buffer);
        }
    }
}

File: java/src/org/openamq/framing/AMQDataBlockEncoder.java
package org.openamq.framing;

import org.apache.log4j.Logger;
import org.apache.mina.common.ByteBuffer;
import org.apache.mina.common.IoSession;
import org.apache.mina.filter.codec.demux.MessageEncoder;
import org.apache.mina.filter.codec.ProtocolEncoderOutput;

import java.util.HashSet;
import java.util.Set;

public class AMQDataBlockEncoder implements MessageEncoder
{
	Logger _logger = Logger.getLogger(AMQDataBlockEncoder.class);

    private Set _messageTypes;

    public AMQDataBlockEncoder()
    {
        _messageTypes = new HashSet();
        _messageTypes.add(EncodableAMQDataBlock.class);
    }

    public void encode(IoSession session, Object message, ProtocolEncoderOutput out) throws Exception
    {
        final AMQDataBlock frame = (AMQDataBlock) message;
        int frameSize = (int)frame.getSize();
        final ByteBuffer buffer = ByteBuffer.allocate(frameSize);
        //buffer.setAutoExpand(true);
        frame.writePayload(buffer);

        if (_logger.isDebugEnabled())
        {
        	_logger.debug("Encoded frame byte-buffer is '" + EncodingUtils.convertToHexString(buffer) + "'");
        }

        buffer.flip();
        out.write(buffer);
    }

    public Set getMessageTypes()
    {
        return _messageTypes;
    }
}
